### PR TITLE
Auto  default placements scenario for shared compute & BYO

### PR DESCRIPTION
**Pull request by kasturinarra**
> Please ask your **team peer** to review your test case first. Once your team peer add `/lgtm` label for your PR, please ask the repo [Maintainers](https://github.com/openshift/openshift-tests-private/blob/master/OWNERS) to have a review and `/approve` it. Thanks!

> Before asking others to **review** your code, please check if your code follows these [rules](https://github.com/openshift/openshift-tests-private/tree/master/test/README.md)
## Your PR Description
Automate validating default placement creation using BYO & SharedCompute
Case: https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitem?id=OCP-52823

### If your run your cases in local, please paste logs here.
<pre>
$ ./bin/extended-platform-tests run all --dry-run|grep <Your-TestCase-ID>|./bin/extended-platform-tests run -f -
[knarra@knarra kcp-tests]$ ./bin/kcp-tests run all --dry-run | grep "Verify default placements" | ./bin/kcp-tests run -f -
started: (0/1/1) "[area/transparent-multi-cluster] Author:knarra-Critical-[Regression] Verify default placements [Suite:kcp/smoke/parallel]"

INFO Oct  4 17:52:56.630: kube-apiserver version: v1.24.3+kcp-v0.8.2-7-g385ef4a7ef73fb
INFO Oct  4 17:52:56.631: Cluster IP family: ipv4
[BeforeEach] [Top Level]
  /home/knarra/automation/Openshift/kcp-tests/test/extended/util/test.go:53
[BeforeEach] [area/transparent-multi-cluster]
  /home/knarra/go/pkg/mod/github.com/!phaow/kubernetes@v0.6.2-kcp/test/e2e/framework/framework.go:154
STEP: Creating a kubernetes client
[BeforeEach] [area/transparent-multi-cluster]
  /home/knarra/automation/Openshift/kcp-tests/test/extended/util/client.go:223
INFO Oct  4 17:52:56.631: Creating workspace "e2e-test-kcp-placements-7tdkk"
INFO Oct  4 17:52:58.959: Workspace "e2e-test-kcp-placements-7tdkk" has been fully provisioned.
[It] Author:knarra-Critical-[Regression] Verify default placements [Suite:kcp/smoke/parallel]
  /home/knarra/automation/Openshift/kcp-tests/test/extended/placements/placements.go:21
STEP: # Validate apibinding creation is successful in user home workspace
STEP: # Verify that default placement has been created for shared compute
STEP: # Verify that APIBinding will be annotated with workload.kcp.dev/skip-default-object-creation
STEP: # Verify creating a namespace in the current workspace
namespace/test52823 created
STEP: # Verify that it randomly select a location and bind to all namespaces in this workspace
STEP: # verify default placement has been created for BYO clusters
INFO Oct  4 17:53:06.673: Creating workspace "e2e-test-kcp-placements-grlkp"
INFO Oct  4 17:53:08.971: Workspace "e2e-test-kcp-placements-grlkp" has been fully provisioned.
INFO Oct  4 17:53:08.971: Creating namespace "e2e-ns-kcp-placements-zs24r"
STEP: # Create workload sync and generate syncer resources manifests in current workspace
STEP: # Apply syncer resources on pcluster and wait for synctarget become ready
namespace/kcp-syncer-mysyncer-ob5jpakf-33gn0ehx created
serviceaccount/kcp-syncer-mysyncer-ob5jpakf-33gn0ehx created
secret/kcp-syncer-mysyncer-ob5jpakf-33gn0ehx-token created
clusterrole.rbac.authorization.k8s.io/kcp-syncer-mysyncer-ob5jpakf-33gn0ehx created
clusterrolebinding.rbac.authorization.k8s.io/kcp-syncer-mysyncer-ob5jpakf-33gn0ehx created
secret/kcp-syncer-mysyncer-ob5jpakf-33gn0ehx created
deployment.apps/kcp-syncer-mysyncer-ob5jpakf-33gn0ehx created
INFO Oct  4 17:53:27.991: SyncTarget/mysyncer-ob5jpakf ready condition is False
INFO Oct  4 17:53:31.392: SyncTarget/mysyncer-ob5jpakf ready condition is True
STEP: # Verify that default placement has been created for BYO cluster
STEP: # Verify that APIBinding will be annotated with workload.kcp.dev/skip-default-object-creation
STEP: # Verify creating a namespace in the current workspace
namespace/test528231 created
STEP: # Verify that it randomly select a location and bind to all namespaces in this workspace
INFO Oct  4 17:53:39.043: Error running /bin/kubectl delete -f /tmp/e2e-test-kcp-placements-grlkp.mysyncer-ob5jpakf.yaml --kubeconfig=/home/knarra/Downloads/kubeconfig_412:
namespace "kcp-syncer-mysyncer-ob5jpakf-33gn0ehx" deleted
serviceaccount "kcp-syncer-mysyncer-ob5jpakf-33gn0ehx" deleted
clusterrole.rbac.authorization.k8s.io "kcp-syncer-mysyncer-ob5jpakf-33gn0ehx" deleted
clusterrolebinding.rbac.authorization.k8s.io "kcp-syncer-mysyncer-ob5jpakf-33gn0ehx" deleted
secret "kcp-syncer-mysyncer-ob5jpakf-33gn0ehx" deleted
deployment.apps "kcp-syncer-mysyncer-ob5jpakf-33gn0ehx" deleted
Error from server (NotFound): error when deleting "/tmp/e2e-test-kcp-placements-grlkp.mysyncer-ob5jpakf.yaml": secrets "kcp-syncer-mysyncer-ob5jpakf-33gn0ehx-token" not found
namespace "kcp-syncer-mysyncer-ob5jpakf-33gn0ehx" deleted
serviceaccount "kcp-syncer-mysyncer-ob5jpakf-33gn0ehx" deleted
clusterrole.rbac.authorization.k8s.io "kcp-syncer-mysyncer-ob5jpakf-33gn0ehx" deleted
clusterrolebinding.rbac.authorization.k8s.io "kcp-syncer-mysyncer-ob5jpakf-33gn0ehx" deleted
secret "kcp-syncer-mysyncer-ob5jpakf-33gn0ehx" deleted
deployment.apps "kcp-syncer-mysyncer-ob5jpakf-33gn0ehx" deleted
Error from server (NotFound): error when deleting "/tmp/e2e-test-kcp-placements-grlkp.mysyncer-ob5jpakf.yaml": secrets "kcp-syncer-mysyncer-ob5jpakf-33gn0ehx-token" not found
synctarget.workload.kcp.dev "mysyncer-ob5jpakf" deleted
[AfterEach] [area/transparent-multi-cluster]
  /home/knarra/automation/Openshift/kcp-tests/test/extended/util/client.go:211
workspace.tenancy.kcp.dev "e2e-test-kcp-placements-7tdkk" deleted
INFO Oct  4 17:53:42.516: Deleted e2e-test-kcp-placements-7tdkk, err: <nil>
workspace.tenancy.kcp.dev "e2e-test-kcp-placements-grlkp" deleted
INFO Oct  4 17:53:44.242: Deleted e2e-test-kcp-placements-grlkp, err: <nil>
[AfterEach] [area/transparent-multi-cluster]
  /home/knarra/go/pkg/mod/github.com/!phaow/kubernetes@v0.6.2-kcp/test/e2e/framework/framework.go:155

passed: (51.8s) 2022-10-04T12:23:44 "[area/transparent-multi-cluster] Author:knarra-Critical-[Regression] Verify default placements [Suite:kcp/smoke/parallel]"

1 pass, 0 skip (51.8s)

</pre>
### If you used the Jenkins job [ginkgo-test](https://mastern-jenkins-csb-openshift-qe.apps.ocp4.prod.psi.redhat.com/job/ocp-common/job/ginkgo-test/) to test your case, please paste the job link here.
<pre>

</pre>

### CODE REVIEW

**Comment by kasturinarra** [comment]
@Phaow @xingxingxia could you please help review ? thanks !!

File: test/extended/placements/placements.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	defer g.GinkgoRecover()
+
+	var (
+		k = exutil.NewCLIWithWorkSpace("kcp-placements")
+	)
+
+	g.It("Author:knarra-Critical-[Regression] Verify default placements", func() {
+		// Shared compute could be only accessed from dev-provided test environments
+		// Skip for non-supported test environments
+		exutil.PreCheckEnvSupport(k, "kcp-stable.apps.kcp-internal", "kcp-unstable.apps.kcp-internal")
```
Comment (xingxingxia) [inline test/extended/placements/placements.go:24]: If this is skipped, then this case is skipped prematurely even if your later BYO code can be run, right? Vice versa. How about using separate "g.It"s for them?

File: test/extended/placements/placements.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+		g.By("# Create workload sync and generate syncer resources manifests in current workspace")
+		mySyncer := nsc.NewSyncTarget()
+		mySyncer.OutputFilePath = "/tmp/" + myWs.Name + "." + mySyncer.Name + ".yaml"
+		mySyncer.Create(k)
+		defer mySyncer.Clean(k)
+
+		g.By("# Apply syncer resources on pcluster and wait for synctarget become ready")
+		defer k.AsPClusterKubeconf().WithoutNamespace().WithoutWorkSpaceServer().Run("delete").Args("-f", mySyncer.OutputFilePath).Execute()
+		err = k.AsPClusterKubeconf().WithoutNamespace().WithoutWorkSpaceServer().Run("apply").Args("-f", mySyncer.OutputFilePath).Execute()
```
Comment (xingxingxia) [inline test/extended/placements/placements.go]: When AsPClusterKubeconf() is used, it does not make sense to involve Workspace server in terms of calling WithoutWorkSpaceServer(). How about updating https://github.com/kcp-dev/kcp-tests/blob/main/test/extended/util/client.go `if !c.withoutWorkSpaceServer` to `if !c.asPClusterKubeconf && !c.withoutWorkSpaceServer` such that you need no call WithoutWorkSpaceServer()?

**Comment by xingxingxia** [comment]
BTW please update your case Status to not stay in "Draft".

File: test/extended/placements/placements.go (hunk 1 above)
Comment (kasturinarra) [reply test/extended/placements/placements.go:24]: Done, addressed.

File: test/extended/placements/placements.go (hunk 2 above)
Comment (kasturinarra) [reply test/extended/placements/placements.go]: Thanks, Done, Addressed !!

**Comment by kasturinarra** [comment]
> BTW please update your case Status to not stay in "Draft".

Done, Now case has been moved to Proposed State.

**Comment by kasturinarra** [comment]
Run logs:
===============
https://privatebin.corp.redhat.com/?a94299b25f740d8d#Cwpadxo142a1pYQmDzepfR5b8i9GpYeX1yrsp6vL69UF

**Comment by xingxingxia** [comment]
LGTM, thank you!

**Pull request merged**